### PR TITLE
fix(api,app): show actual error reason on invite acceptance page

### PR DESCRIPTION
**Pull request by faizkhairi**
## Description

Fixes #26699

The accept-invite page showed a generic error code translation (e.g., "Invalid payload") instead of the actual API error reason when invite acceptance failed. This is because `translateAPIError()` only extracts the error `code` and maps it to an i18n string, discarding the human-readable `reason` field entirely.

### Previous Attempts

Five previous PRs attempted to fix this issue:

- #26708, #26769, #26950 -- changed the backend error wording (rejected: still leaked account state)
- #26846 -- introduced a new error code `INVITE_INVALID` (rejected: over-engineered, wrong layer)
- #26962 -- made the error generic (rejected per [feedback from @AlexGaillard](https://github.com/directus/directus/pull/26962#issuecomment-3102168182): the fix needs **both frontend and backend changes** since the error was not rendering on the frontend at all)

All five PRs only touched the backend. None addressed the frontend, which is where the error display breaks.

### Root Cause

`translateAPIError()` in `app/src/lang/index.ts` extracts `extensions.code` from the API error and looks up a generic i18n translation. For `INVALID_PAYLOAD`, this returns "Invalid payload". The actual reason ("This invite is no longer valid") lives in `extensions.reason` but is never shown to the user.

### The Fix

**Backend** (`api/src/services/users.ts`):
- Replace the email-leaking error `Email address ${email} hasn't been invited` with a generic `This invite is no longer valid` to prevent account enumeration

**Frontend** (`app/src/utils/get-error-reason.ts` + `app/src/routes/accept-invite.vue`):
- Extract the error formatting logic into a reusable `getErrorReason()` utility (mirrors the existing `extract-error-code.ts` pattern)
- Prefers `extensions.reason` from the API error response when available
- Falls back to `translateAPIError()` for errors without a `reason` field

This follows the same per-page error customization pattern used in `app/src/routes/register/register-form.vue` (lines 37-39).

### What Changes

| Scenario | Before | After |
|----------|--------|-------|
| Invite invalid (user already active) | "Invalid payload" | "This invite is no longer valid" |
| Password too weak | "Validation failed" | "Validation failed" (unchanged, no reason field) |
| Token expired | "Token expired" | "Token expired" (unchanged) |

### Changes

| File | Change |
|------|--------|
| `api/src/services/users.ts` | Generic error message (1 line) |
| `app/src/utils/get-error-reason.ts` | Extracted error reason utility function |
| `app/src/routes/accept-invite.vue` | Use `getErrorReason()` for error display |
| `api/src/services/users.test.ts` | Tests for acceptInvite error handling (catch pattern + reason assertion) |
| `app/src/utils/get-error-reason.test.ts` | Unit tests for both branches + empty errors edge case |
| `.changeset/fix-invite-error-display.md` | Changeset for `@directus/api` + `@directus/app` |

### CODE REVIEW

File: api/src/services/users.test.ts [hunk 1]
```diff
@@ -439,5 +448,43 @@ describe('Integration Tests', () => {
...
+				});
+
+				const service = new UsersService({
+					knex: db,
+					schema,
+				});
+
+				const promise = service.acceptInvite('fake-token', 'Password123!');
+
+				await expect(promise).rejects.toThrow(InvalidPayloadError);
```
Comment (LZylstra) [inline api/src/services/users.test.ts]: The test checks the email is absent but never pins the actual message the user will see. Worth asserting the specific string so a future refactor that swaps in a different generic message doesn't silently pass:

```ts
const err = await promise.catch(e => e);
expect(err.extensions?.reason).toBe('This invite is no longer valid');
```

File: api/src/services/users.test.ts [hunk 2]
```diff
@@ -439,5 +448,43 @@ describe('Integration Tests', () => {
...
+
+				await expect(promise).rejects.toThrow(
+					expect.not.objectContaining({
+						message: expect.stringContaining('test@example.com'),
+					}),
+				);
```
Comment (LZylstra) [inline api/src/services/users.test.ts]: Reusing the same promise for two `rejects` assertions works in Vitest (the second matcher does enforce), but the double-await pattern is a bit confusing to read. If you restructure to catch the error directly it also gives you a clean place to add the `extensions.reason` check:

```ts
const err = await promise.catch(e => e);
expect(err).toBeInstanceOf(InvalidPayloadError);
expect(err.message).not.toContain('test@example.com');
expect(err.extensions?.reason).toBe('This invite is no longer valid');
```

File: app/src/routes/accept-invite.vue [hunk 3]
```diff
@@ -26,6 +26,8 @@ const done = ref(false);
 
 const errorFormatted = computed(() => {
 	if (error.value) {
+		const reason = error.value?.response?.data?.errors?.[0]?.extensions?.reason;
+		if (reason) return reason;
```
Comment (LZylstra) [inline app/src/routes/accept-invite.vue]: The two new branches here (`reason` present vs. fallback to `translateAPIError`) aren't covered by any tests, if this regresses it'll be silent. Easiest path would be extracting the logic into a small pure function and unit testing that, but even a quick component test with a mocked error response would do the job.

**Review by LZylstra — CHANGES_REQUESTED** [review]

**Comment by LZylstra** [comment]
Nice work getting both layers fixed. The security fix is right, the fallback behavior is clean, and the changeset is in good shape.

Two small test things worth tightening before merge, plus one recommendation on the frontend side.

**Comment by faizkhairi** [comment]
> Nice work getting both layers fixed. The security fix is right, the fallback behavior is clean, and the changeset is in good shape.
> 
> Two small test things worth tightening before merge, plus one recommendation on the frontend side.

Understood, thanks for the input. Greatly appreciates it @LZylstra 👍🏼 

Will do the CR now.

File: api/src/services/users.test.ts (hunk 1 above)
Comment (faizkhairi) [reply api/src/services/users.test.ts]: Good call. Added the extensions.reason assertion to pin the specific message.

File: api/src/services/users.test.ts (hunk 2 above)
Comment (faizkhairi) [reply api/src/services/users.test.ts]: Agreed, refactored to the `.catch(e => e)` pattern. Much cleaner with all three assertions on the same `err` object.

File: app/src/routes/accept-invite.vue (hunk 3 above)
Comment (faizkhairi) [reply app/src/routes/accept-invite.vue]: Went with the pure function approach. Extracted `getErrorReason()` into `app/src/utils/get-error-reason.ts` (mirrors the existing `extract-error-code.ts` pattern) with 3 unit tests covering: reason present, reason absent (translateAPIError fallback), and empty errors array.

**Comment by faizkhairi** [comment]
@LZylstra All three review points addressed in 6c3e978:

1. **Pin error message**: Added `expect(err.extensions?.reason).toBe('This invite is no longer valid')` 
2. **Catch pattern**: Refactored to `.catch(e => e)` with three clean assertions on the same error object
3. **Frontend test coverage**: Extracted branching logic into `getErrorReason()` utility (`app/src/utils/get-error-reason.ts`) with 3 unit tests covering both branches + empty errors edge case

Ready for re-review when you get a chance.

**Review by LZylstra — APPROVED** [review]
LGTM

File: .changeset/fix-invite-error-display.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@directus/api": patch
+"@directus/app": patch
+---
+
+Fixed invite acceptance to show the actual error reason on the frontend instead of a generic "Invalid payload" message. Also prevents email address leakage in the error response.
```
Comment (AlexGaillard) [inline .changeset/fix-invite-error-display.md]:
```suggestion
Fixed invite acceptance error to display correctly on the frontend and remove email leakage
```

**Comment by faizkhairi** [comment]
Rebased on latest main to resolve merge conflicts. All CI checks passing locally.

**Comment by faizkhairi** [comment]
The Unit Tests failure is unrelated to this PR -- it's a date-sensitive test in `v-date-picker.test.ts:909`:

```
v-date-picker > month and year selection > clamps day when changing year affects leap year
AssertionError: expected 29 to be less than or equal to 28
```

This test passed on other PRs merged days earlier (#26986, Mar 27) and fails consistently now regardless of re-runs. None of our changes touch `v-date-picker`. The second error (`document is not defined` in `overlay-item.test.ts`) is also a pre-existing teardown issue with `universal-cookie`.

All other checks (Format, Lint, Stylelint, Changeset) pass. Rebased on latest main.

**Review by AlexGaillard — APPROVED** [review]
Thanks for the fix @faizkhairi 👍

**Comment by faizkhairi** [comment]
> Thanks for the fix @faizkhairi 👍

all goood~ 🚀

**Comment by AlexGaillard** [comment]
Actually, I'm realizing now that this doesn't account for translations 🤔. While it's arguable that this is an improvement of the current approach, that at least does provide translations for the error in different languages. I think we're going to have to rethink this approach.

**Comment by faizkhairi** [comment]
Fair point, `extensions.reason` is English-only and bypasses the i18n system entirely. I see a few paths:

1. Extend `getErrorReason()` to normalize the reason string to an i18n key and check for a translation before falling back to the raw string, then add the specific key to `en-US.yaml` for Crowdin to pick up
2. Add a dedicated error code for the invite-invalid case (revisiting the `INVITE_INVALID` approach from #26846, now that the frontend display is also addressed)

Which direction fits better with how you'd like this handled?

**Comment by AlexGaillard** [comment]
Number 2 problably

File: .changeset/fix-invite-error-display.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+"@directus/api": patch
+"@directus/app": patch
+"@directus/errors": patch
+---
+
+Introduce a dedicated `INVALID_INVITE` error code for expired/invalid invite acceptance. The error message is now translated via the i18n system on the frontend instead of being read from `extensions.reason`, ensuring it is localizable and consistent with other error messages. Also removes the `getErrorReason` utility which bypassed the translation system.
```
Comment (AlexGaillard) [inline .changeset/fix-invite-error-display.md]:
```suggestion
Fixed invite acceptance error to display correctly on the frontend and allow for error translation
```

This is fine how it was before, don't need this much detail

**Review by AlexGaillard — CHANGES_REQUESTED** [review]

**Review by AlexGaillard — APPROVED** [review]
Tested with other translations and works nicely 👍 Took a while but we got there in the end, TY @faizkhairi
